File: apps/mobile/src/features/core/native-core-adapters/feedback.ts
import type { FeedbackEvent } from '@glimpse/shared';

import { nativeCoreBridgeHelpers } from '../native-core-bridge';
import type { BridgeCoreClient } from '../types';

import type { NativeAdapterDeps } from './common';

const { fromNitroFeedbackEvent, toNitroFeedbackEvent } = nativeCoreBridgeHelpers;

export function createFeedbackAdapter(
  deps: NativeAdapterDeps,
): Pick<BridgeCoreClient, 'listRecentFeedbackEvents' | 'logRecommendationFeedback'> {
  const { fallbackClient, runCoreAsync } = deps;

  return {
    async listRecentFeedbackEvents(limit: number): Promise<FeedbackEvent[]> {
      return runCoreAsync(
        async (client) =>
          (await client.listRecentFeedbackEvents(limit)).map(fromNitroFeedbackEvent),
        () => fallbackClient.listRecentFeedbackEvents(limit),
      );
    },

    async logRecommendationFeedback(event: FeedbackEvent): Promise<FeedbackEvent> {
      return runCoreAsync(
        async (client) =>
          fromNitroFeedbackEvent(
            await client.logRecommendationFeedback(toNitroFeedbackEvent(event)),
          ),
        () => fallbackClient.logRecommendationFeedback(event),
      );
    },
  };
}

File: packages/core-rust/src/application/feedback.rs
use crate::error::Result;
use crate::models::FeedbackEvent;

use super::SharedCore;

impl SharedCore {
    pub fn list_recent_feedback_events(&self, limit: usize) -> Result<Vec<FeedbackEvent>> {
        self.client().list_recent_feedback_events(limit)
    }

    pub fn log_recommendation_feedback(&self, event: &FeedbackEvent) -> Result<FeedbackEvent> {
        self.client().log_recommendation_feedback(event)
    }
}
